pkg/optimizer/datapath_csp.go

```diff
@@ -0,0 +1,700 @@
+
+	// Finally, make sure a storage account is assigned iff there is a sink interface and it is non-virtual
+	if dpc.problemData.Context.Flow == taxonomy.WriteFlow && !dpc.problemData.Context.Requirements.FlowParams.IsNewDataSet {
+		return // no need to allocate storage, write destination is known
+	}
+
+	noSaRequiredModCaps := []string{}
+	for modCapIdx, modCap := range dpc.modulesCapabilities {
+		if modCap.virtualSink || !modCap.hasSink {
+			noSaRequiredModCaps = append(noSaRequiredModCaps, strconv.Itoa(modCapIdx+1))
+		}
+	}
+	noSaRequiredVarName := dpc.addSetInIndicator(modCapVarname, noSaRequiredModCaps, pathLength)
+	realSA := dpc.addEqualityIndicator(saVarname, 0, pathLength, false) // a value >0 means a storage account is allocated
+	for pathPos := 1; pathPos <= pathLength; pathPos++ {
+		noSaRequiredAtPos := varAtPos(noSaRequiredVarName, pathPos)
+		realSAAtPos := varAtPos(realSA, pathPos)
+		dpc.fzModel.AddConstraint(BoolNotEqConstraint, []string{realSAAtPos, noSaRequiredAtPos})
+	}
+}
+
+// Return a list of indexes of interfaces that match the input interface
+func (dpc *DataPathCSP) getMatchingInterfaces(refIntfc *taxonomy.Interface) []string {
+	res := []string{}
+	for intfc, intfcIdx := range dpc.interfaceIdx {
+		if interfacesMatch(refIntfc, &intfc) {
+			res = append(res, strconv.Itoa(intfcIdx))
+		}
+	}
+	return res
+}
+
+// Add constraints to ensure interface selection matches module-capability selection
+func (dpc *DataPathCSP) modCapSupportsIntfc(pathLength int) {
+	for intfc, intfcIdx := range dpc.interfaceIdx {
+		for modCapIdx, modCap := range dpc.modulesCapabilities {
+			modcapSupportsIntfcSrc := false
+			modcapSupportsIntfcSink := false
+			for _, modifc := range modCap.capability.SupportedInterfaces {
+				modcapSupportsIntfcSrc = modcapSupportsIntfcSrc || interfacesMatch(modifc.Source, &intfc)
+				modcapSupportsIntfcSink = modcapSupportsIntfcSink || interfacesMatch(modifc.Sink, &intfc)
+			}
+			if modCap.virtualSource || modCap.virtualSink {
+				capAPI := modCap.capability.API
+				apiIntfc := &taxonomy.Interface{Protocol: capAPI.Connection.Name, DataFormat: capAPI.DataFormat}
+				modcapSupportsIntfcSrc = modcapSupportsIntfcSrc || modCap.virtualSource && interfacesMatch(apiIntfc, &intfc)
+				modcapSupportsIntfcSink = modcapSupportsIntfcSink || modCap.virtualSink && interfacesMatch(apiIntfc, &intfc)
+			}
+			if !modcapSupportsIntfcSrc {
+				dpc.preventAssignments([]string{modCapVarname, srcIntfcVarname}, []int{modCapIdx + 1, intfcIdx}, pathLength)
+			}
+			if !modcapSupportsIntfcSink {
+				dpc.preventAssignments([]string{modCapVarname, sinkIntfcVarname}, []int{modCapIdx + 1, intfcIdx}, pathLength)
+			}
+		}
+	}
+}
+
+// If there are optimization goals set, defines appropriate variables and sets the CSP-solver optimization goal
+// Otherwise, just sets the CSP-solver goal as "satisfy"
+func (dpc *DataPathCSP) addOptimizationGoals(pathLength int) error {
```

Comment:
What in the current code prevents us from optimizing for more than one optimization goal?

Reply:
Do you mean "bandwidth-like" optimization goals?
Such goals will return an empty InstanceType when calling `AttributeManager.GetInstanceType(attr)` and will be ignored.